internal/k8sutil/k8sutil.go

```diff
@@ -150,21 +153,56 @@ func QuoteAttestationDelete(ctx context.Context, c client.Client, instanceName s
 		},
 	}
 
-	err := c.Delete(ctx, sgxAttestation)
-	if err != nil && errors.IsNotFound(err) {
-		return nil
+	if err := client.IgnoreNotFound(UnsetFinalizer(ctx, c, sgxAttestation)); err != nil {
+		return fmt.Errorf("failed unset finalizer for '%s/%s': %v", ns, instanceName, err)
 	}
-	return err
+
+	return client.IgnoreNotFound(c.Delete(ctx, sgxAttestation))
 }
 
-// Converts signer name to valid Kubernetes object name
-//  Ex:- intel.com/tcs -> tcs.intel.com
-//       tcsissuer.tcs.intel.com/sgx-ca1 -> sgx-ca1.tcsissuer.intel.tcs.com
-func SignerNameToResourceName(signerName string) string {
+// Converts signer name to valid Kubernetes object name and nanespace
+//  Ex:- intel.com/tcs -> tcs.intel.com, ""
+///      tcsissuer.tcs.intel.com/sandbox.sgx-ca -> sgx-ca.tcs.intel.com, sandbox
+//       tcsclusterissuer.tcs.intel.com/sgx-ca1 -> sgx-ca1.tcsclusterissuer.intel.tcs.com, ""
+func SignerNameToResourceNameAndNamespace(signerName string) (string, string) {
 	slices := strings.SplitN(signerName, "/", 2)
 	if len(slices) == 2 {
-		return slices[1] + "." + slices[0]
+		nameParts := strings.SplitN(slices[1], ".", 2)
+		if len(nameParts) == 2 {
+			return nameParts[1] + "." + slices[0], nameParts[0]
+		}
+		return slices[1] + "." + slices[0], ""
 	}
 
-	return slices[0]
+	return slices[0], ""
+}
+
+func UnsetFinalizer(ctx context.Context, c client.Client, obj client.Object) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(2*time.Minute))
+	defer cancel()
+	key := client.ObjectKeyFromObject(obj)
+	err := c.Get(timeoutCtx, key, obj)
+	if err != nil {
+		return err
+	}
+
+	list := obj.GetFinalizers()
+	found := false
+	for i, finalizer := range list {
+		if finalizer == TCSFinalizer {
+			found = true
+			list = append(list[:i], list[i+1:]...)
+			break
+		}
+	}
+
+	if found {
+		obj.SetFinalizers(list)
+		timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(2*time.Minute))
+		defer cancel()
+		if err := client.IgnoreNotFound(c.Update(timeoutCtx, obj)); err != nil {
```

Comment:
Because there are multiple finalizers it's assumed someone else other than TCS can set them? But does calling `update()` cause a race if someone else is setting them at the same time?

Reply:
Okay,  I will move it to patch(). that might address the issue.

Reply:
Done,